ethcore/src/client/client.rs

```diff
@@ -171,6 +171,35 @@ struct Importer {
 	pub bad_blocks: bad_blocks::BadBlocks,
 }
 
+
+/// Newtype around an `Arc<BlockChain>` so we can glue together `BlockChain` (from the `blockchain`
+/// crate) and `DatabaseExtras` (from the `trace` crate).
+struct BlockChainWithExtras(Arc<BlockChain>);
+
+impl trace::DatabaseExtras for BlockChainWithExtras {
+	fn block_hash(&self, block_number: BlockNumber) -> Option<H256> {
+		(self.0.as_ref() as &dyn BlockProvider).block_hash(block_number)
+	}
+
+	fn transaction_hash(&self, block_number: BlockNumber, tx_position: usize) -> Option<H256> {
+		(self.0.as_ref() as &dyn BlockProvider).block_hash(block_number)
```

Comment:
```suggestion
		BlockProvider::block_hash(&self.0, block_number)
```

Reply:
```
error[E0277]: the trait bound `std::sync::Arc<blockchain::BlockChain>: blockchain::BlockProvider` is not satisfied
   --> ethcore/src/client/client.rs:181:3
    |
181 |         BlockProvider::block_hash(&self.0, block_number)
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `blockchain::BlockProvider` is not implemented for `std::sync::Arc<blockchain::BlockChain>`
    |
    = note: required by `blockchain::BlockProvider::block_hash`
```

Reply:
I guess `self.0` is of type `std::sync::Arc<blockchain::BlockChain>` you'll need to deref.

`BlockProvider::block_hash(&*self.0, block_number)`